File: tests/pyrtl/common/ram.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pyrtl
pyrtl.set_debug_mode(debug=True)

DATA_WIDTH = 8
ADDR_WIDTH = 8
RAM_DEPTH = 1 << 8

address = pyrtl.Input(ADDR_WIDTH, "address")
cs = pyrtl.Input(1, "cs")
we = pyrtl.Input(1, "we")
oe = pyrtl.Input(1, "oe")
data_in = pyrtl.Input(DATA_WIDTH, "data_in")
data_out = pyrtl.Output(DATA_WIDTH, "data_out")

mem = pyrtl.memory.MemBlock(RAM_DEPTH, ADDR_WIDTH,
                            name="mem", asynchronous=True)
mem[address] <<= pyrtl.MemBlock.EnabledWrite(data_in, we & cs)

data_out_wire = pyrtl.wire.WireVector(DATA_WIDTH, "DATA_WIDTH")

with pyrtl.conditional_assignment:
    with oe:
        with we:
            pass
        with pyrtl.otherwise:
            with cs:
                data_out_wire |= mem[address]

data_out <<= data_out_wire

# Setup the simulation
sim_trace = pyrtl.SimulationTrace()
sim = pyrtl.FastSimulation(tracer=sim_trace, memory_value_map={mem: {0: 0}})

File: tests/chisel/sock/ram.scala
package ram

import chisel3._

class ram extends Module {
    val DATA_WIDTH: Int = 8
    val ADDR_WIDTH: Int = 8
    val RAM_DEPTH: Int = 1 << 8
    val io = IO(new Bundle {
        val address  = Input(UInt(ADDR_WIDTH.W))
        val cs       = Input(Bool())
        val we       = Input(Bool())
        val oe       = Input(Bool())
        val data_in  = Input(UInt(DATA_WIDTH.W))
        val data_out = Output(UInt(DATA_WIDTH.W))
    })

    val data_out_wire = Reg(UInt(DATA_WIDTH.W))
    val mem = Mem(RAM_DEPTH, UInt(DATA_WIDTH.W))
    data_out_wire := DontCare

    when (io.cs) {
        when(!io.we && io.oe) {
            // read
            data_out_wire := mem.read(io.address)
        }
        .otherwise {
            // write
            mem.write(io.address, io.data_in)
        }
    }

    io.data_out := data_out_wire

}
